contracts/strategies/StrategyDAI3poolIron.sol

```diff
@@ -0,0 +1,562 @@
+    }
+
+    function deposit() public isAuthorized {
+        rebalance();
+        uint256 _want = (IERC20(want).balanceOf(address(this))).sub(tank);
+        if (_want > 0) {
+            IERC20(want).safeApprove(_3pool, 0);
+            IERC20(want).safeApprove(_3pool, _want);
+            uint256 v = _want.mul(1e18).div(ICurveFi(_3pool).get_virtual_price());
+            ICurveFi(_3pool).add_liquidity([_want, 0, 0], v.mul(DENOMINATOR.sub(slip)).div(DENOMINATOR));
+        }
+        uint256 _bal = IERC20(_3crv).balanceOf(address(this));
+        if (_bal > 0) {
+            IERC20(_3crv).safeApprove(y3crv, 0);
+            IERC20(_3crv).safeApprove(y3crv, _bal);
+            yvERC20(y3crv).deposit(_bal);
+        }
+        _bal = IERC20(y3crv).balanceOf(address(this));
+        if (_bal > 0) {
+            IERC20(y3crv).safeApprove(cToken, 0);
+            IERC20(y3crv).safeApprove(cToken, _bal);
+            require(ICErc20(cToken).mint(_bal) == 0, "!deposit");
+            urn = ICErc20(cToken).balanceOfUnderlying(address(this));
+        }
+    }
+
+    // Controller only function for creating additional rewards from dust
+    function withdraw(IERC20 _asset) external returns (uint256 balance) {
+        require(msg.sender == controller, "!controller");
+        require(want != address(_asset), "want");
+        require(_3crv != address(_asset), "3crv");
+        require(y3crv != address(_asset), "y3crv");
+        require(cToken != address(_asset), "cyY3CRV");
+        balance = _asset.balanceOf(address(this));
+        _asset.safeTransfer(controller, balance);
+    }
+
+    // Withdraw partial funds, normally used with a vault withdrawal
+    function withdraw(uint256 _amount) external {
+        require(msg.sender == controller, "!controller");
+
+        rebalance();
+        uint256 _balance = IERC20(want).balanceOf(address(this));
+        if (_balance < _amount) {
+            _amount = _withdrawSome(_amount.sub(_balance));
+            _amount = _amount.add(_balance);
+            tank = 0;
+        } else {
+            if (tank >= _amount) tank = tank.sub(_amount);
+            else tank = 0;
+        }
+
+        address _vault = Controller(controller).vaults(address(want));
+        require(_vault != address(0), "!vault"); // additional protection so we don't burn the funds
+        uint256 _fee = _amount.mul(withdrawalFee).div(DENOMINATOR);
+        IERC20(want).safeTransfer(Controller(controller).rewards(), _fee);
+        IERC20(want).safeTransfer(_vault, _amount.sub(_fee));
+    }
+
+    function _withdrawSome(uint256 _amount) internal returns (uint256) {
+        uint256 _amnt = _amount.mul(1e18).div(ICurveFi(_3pool).get_virtual_price());
+        uint256 _amt = _amnt.mul(1e18).div(yvERC20(y3crv).getPricePerFullShare());
+        uint256 _bal = ICErc20(cToken).balanceOfUnderlying(address(this));
+        if (_amt > _bal) _amt = _bal;
+        require(ICErc20(cToken).redeemUnderlying(_amt) == 0, "!withdrawSome");
```

Comment:
Not fully familiar with iron bank token, does `reedemUnderlying` fail on zero amount argument and if that is the case, is there a chance we are ever passing zero?

Reply:
successful on zero amount, and returns 0 as `NO_ERROR`.
I've tested this case when migrating this strategy to a new strategy. All underlying y3crv will be migrated to the new strategy, and then `withdrawAll` will be called.